doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,352 @@
+  unlike LSM-tree DB (proposed design is similar to LSM-tree which has only level-0)
+  
+
+With NVM
+--------
+With NVM, we are able to make use of a different scheme to store the data depending replication
+performance.
+All of the data store NVM first with out-of-place update manner, then flush them to NVMe storage device
+using in-place update scheme. This is feasible when replication latency (> 0.5 ms) is slower than
+NVMe storage performance (< 0.5 ms) because flush can store all buffered data in NVM in batch-manner.
+
+
+Detailed Design 
+===============
+
+
+- Onode lookup
+Upon onode lookup requests
+First, we hash CID (collection ID) to get the hash index of the hash table. Note that the entire hash table can be 
+pre-allocated when we limit the maximum number of objects that can be created in an AG. With the hash index, 
+the buckets are scanned until it finds the hash entry for the target object ID.
+A hash entry contains <CID, OID, Type, Offset, Segment>.
+With the hash entry, we can find the disk location of the target onode.
+Since the OID meta entry in the segment contains <OID, Chunk Number, (Offset, len), (Offset, len), … >, 
+we can read the onode from the target chunk.
+
+- Allocation
+Allocation Groups
+Entire disk space is divided into a number of equally sized chunks called Allocation Groups (AG).
+Each reactor thread in crimson-osd is responsible for a set of AGs (# of AGs / # of threads).
+Each AG has its own data structures to manage the disk partition.
+
+- Data allocation
+As we explained above, the entire hash table is pre-allocated in each AG. Given CID and OID, we can map any data to 
+the segment and offset. A segment is comprised of contiguous blocks. Blocks can be allocated by searching the free space 
+tracking data structure (we explain below).
+Based on onode lookup, we probably know where OID is located via segment number. And, segment contains multiple entries 
+such as oid meta entries (object_info_t, data etc.), small data, and data. (segment header describe what entries exists there).
+So a single write can store a transaction data at once.
+
+- Free space tracking
+We can use either data block bitmap in EXT4 or extent-based B+tree in XFS for free space tracking.
+Our first prototype will be implemented based on the block bitmap for the sake of implementation.
+Regarding allocation bitmap, we have a plan to re-use the existing the design of bitmapfreelistmanager and bitmapallocator 
+in Bluestore (Underlying storage should be changed KV db to raw device)
+
+- Omap
+In this design (see below figure), omap is not different from data and onode. They all can be retrieved via hash table lookup
+
+- Fragmentation
+Internal fragmentation
+We pack different type of data/metadata in a single block as many as possible to reduce internal fragmentation.
+Extent-based B+tree may help reduce this further by allocating contiguous blocks that best fit for the object
```

Comment:
That's not the question.  Over time, data packed into your blocks will become obsolete and those blocks will contain "free" space that can't be used for allocation.  Is the plan to have a background process to reclaim under-used blocks?